ios/sound.mm

```diff
@@ -24,209 +24,277 @@
-  return strStat;
-   */
-    return "";
+    QMutexLocker locker ( &pSound->MutexAudioProcessCallback );
+    Float32*     pData = static_cast<Float32*> ( inInputData->mBuffers[0].mData );
+
+    // copy input data
+    for ( int i = 0; i < pSound->iCoreAudioBufferSizeMono; i++ )
+    {
+        // copy left and right channels separately
+        pSound->vecsTmpAudioSndCrdStereo[2 * i]     = (short) ( pData[2 * i] * _MAXSHORT );     // left
+        pSound->vecsTmpAudioSndCrdStereo[2 * i + 1] = (short) ( pData[2 * i + 1] * _MAXSHORT ); // right
+    }
+    pSound->ProcessCallback ( pSound->vecsTmpAudioSndCrdStereo );
 }
 
-QString CSound::CheckDeviceCapabilities ( const int iDriverIdx )
+/* Implementation *************************************************************/
+CSound::CSound ( void ( *fpNewProcessCallback ) ( CVector<short>& psData, void* arg ),
+                 void*          arg,
+                 const QString& strMIDISetup,
+                 const bool,
+                 const QString& ) :
+    CSoundBase ( "CoreAudio iOS", fpNewProcessCallback, arg, strMIDISetup )
 {
-  /*// This function checks if our required input/output channel
-  // properties are supported by the selected device. If the return
-  // string is empty, the device can be used, otherwise the error
-  // message is returned.
-
-  // check sample rate of input
-  audioSession.setPreferredSampleRate( SYSTEM_SAMPLE_RATE_HZ );
-  // if false, try to set it
-  if ( audioSession.sampleRate != SYSTEM_SAMPLE_RATE_HZ )
-  {
-    throw CGenErr ( tr ("Could not set sample rate. Please close other applications playing audio.") );
-  }
-
-  // special case with 4 input channels: support adding channels
-    if ( ( lNumInChan == 4 ) && bInputChMixingSupported )
+    try
     {
-        // add four mixed channels (i.e. 4 normal, 4 mixed channels)
-        lNumInChanPlusAddChan = 8;
-
-        for ( int iCh = 0; iCh < lNumInChanPlusAddChan; iCh++ )
-        {
-            int iSelCH, iSelAddCH;
-
-            GetSelCHAndAddCH ( iCh, lNumInChan, iSelCH, iSelAddCH );
-
-            if ( iSelAddCH >= 0 )
-            {
-                // for mixed channels, show both audio channel names to be mixed
-                channelInputName[iCh] =
-                    channelInputName[iSelCH] + " + " + channelInputName[iSelAddCH];
-            }
-        }
+        NSError* audioSessionError = nil;
+
+        [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayAndRecord error:&audioSessionError];
+        [[AVAudioSession sharedInstance] requestRecordPermission:^( BOOL granted ) {
+          if ( granted )
+          {
+              // ok
+          }
+        }];
+        [[AVAudioSession sharedInstance] setMode:AVAudioSessionModeMeasurement error:&audioSessionError];
     }
-    else
+    catch ( const CGenErr& generr )
     {
-        // regular case: no mixing input channels used
-        lNumInChanPlusAddChan = lNumInChan;
+        qDebug ( "Sound exception ...." );
```

Comment:
Generally I've seen this style in the code:
```
qDebug() << "Sound exception";
```
Is there a reason to do it differently?

Reply:
Just 2 ways to do the same thing. Although your way requires #include

Reply:
That's why we've code style guidelines that say "do it like it's done elsewhere" rather than "do whatever you like".

Reply:
I actually have a valid reason to limit the libraries included. Initially, the iOS code had a few more unnecessary #includes, and @ann0see iPhone couldn't even launch Jamulus because it took too long. Launched after removing libs, so I prefer to stay away from #include when possible.

The `qDebug() <<` syntax requires `#include <QDebug>`.

Reply:
#include changes don't change the launch time.

If the function call requires the library to work then I don't see how you can avoid it.

As far as start up performance goes, you have to not block the main thread for long periods or you'll be killed by the system. It's pretty established that this is the ::init call happening too many times.

Reply:
> #include changes don't change the launch time.

It seemed to do though. We did a test with @ann0see iphone which at first couldn't launch because it took too long. Removed some unused libs, and it launched.
 
> If the function call requires the library to work then I don't see how you can avoid it.

The thing here is qDebug("something") can be called without `#include QDebug`, while `qDebug() << "something"` can't. Although I'm sure QDebug is already included in other ios-unrelated parts of the code, I'm just trying to optimize because ios launch time is yet to be improved.

> As far as start up performance goes, you have to not block the main thread for long periods or you'll be killed by the system. It's pretty established that this is the ::init call happening too many times.

Yeah I proposed multithreading...

I actually see another potential issue: MAX_NUM_CHANNELS (default value=150). If I'm not mistaken, at launch, Jamulus creates MAX_NUM_CHANNELS faders in the mixer, and hide unused ones. 

However, seeing that other OSes don't seem to have problem with this, the problem might be CSound::Init, which is specific to each system.